ui/test/nuxeo-document-layout.test.js

```diff
@@ -0,0 +1,232 @@
+
+Nuxeo = Nuxeo || {};
+Nuxeo.LayoutBehavior = LayoutBehavior;
+
+/* eslint-disable no-unused-expressions */
+suite('<nuxeo-document-layout>', () => {
+  let documentLayout;
+
+  const buildDoc = (props) => {
+    const doc = { type: 'Test', uid: '12ae', properties: {} };
+    Object.assign(doc.properties, props);
+    return doc;
+  };
+
+  const buildLayout = async (doc = buildDoc(), layout = 'edit') => {
+    const dl = await fixture(
+      html`
+        <nuxeo-document-layout .document="${doc}" layout="${layout}"></nuxeo-document-layout>
+      `,
+    );
+    if (!dl.element) {
+      await Promise.race([
+        waitForEvent(dl.$.layout, 'element-changed'),
+        waitForAttrMutation(dl.$.layout.$.error, 'hidden', null),
+      ]);
+    }
+    await flush();
+    return dl;
+  };
+
+  const getWidget = (label) => {
+    if (!documentLayout || !documentLayout.element) {
+      return null;
+    }
+    return documentLayout.element.shadowRoot.querySelector(`[role="widget"][label="${label}"]`);
+  };
+
+  test('Should display an error when the layout is not found', async () => {
+    documentLayout = await buildLayout(buildDoc(), 'view'); // we have no such view layout, so it will result in 404
+    expect(documentLayout.$.layout.$.error.hidden).to.be.false;
+    expect(documentLayout.$.layout.$.error.message).to.equal('documentView.layoutNotFound');
+    isElementVisible(documentLayout.$.layout.$.error);
+    expect(documentLayout.element).to.be.undefined;
+  });
+
+  test('Should load a layout', async () => {
+    documentLayout = await buildLayout();
+    // check the layout was loaded
+    expect(documentLayout.element).to.not.be.null;
+    expect(documentLayout.element.tagName).to.equal('NUXEO-TEST-EDIT-LAYOUT');
+    // assert layout integrity
+    expect(documentLayout.element.document).to.equal(documentLayout.document);
+    const { children } = documentLayout.element.shadowRoot;
+    expect(children).to.have.lengthOf(3);
+    expect(children[0].tagName).to.equal('STYLE');
+    expect(children[1].tagName).to.equal('NUXEO-INPUT');
+    expect(children[2].tagName).to.equal('NUXEO-TEXTAREA');
+  });
+
+  test('Should fire a "document-layout-changed" event when the layout is loaded', async () => {
+    const ePromise = new Promise((resolve) => {
+      document.addEventListener('document-layout-changed', (e) => resolve(e));
+    });
+    documentLayout = await buildLayout();
+    // check that the event was fired from the document layout
+    const event = await ePromise;
+    expect(event.target).to.equal(documentLayout);
+    // ensure that the event bubbles and is composed
+    expect(event.bubbles).to.be.true;
+    expect(event.composed).to.be.true;
+    // verify event details
+    expect(event.detail).to.exist;
+    expect(event.detail.element).to.equal(documentLayout.element);
+    expect(event.detail.layout).to.equal(documentLayout.layout);
+  });
+
+  test('Should do input validation', async () => {
```

Comment:
Can't we define a `when`? Maybe it could be a sign that we can break this in a specific suite for validations (just an example!) and we can break the tests: when it is not expected to be valid and when it is.

Reply:
We can just say "when the validate method is called", which is what's happening. WDYT?